core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -48,20 +123,85 @@ import Pull._
+    *
+    * - If `this` pull completes successfully with a result of type R, 
+    *   the `f` function is applied to the result to build a new Pull, 
+    *   and compilation continues by running that pull. The end return of 
+    *   the composed pull is that of the generated pull.
+    * - If this pull fails or is interrupted, the failure of interruption is 
+    *   the returned of the composed pull.
+    *
+    * The composed pull will first emit all the outputs emitted by `this` pull,
+    * and if successful will start emitting the outputs from the generated pull.
+    *
+    * This operation does not modify resource scope boundaries. The generated `post` 
+    * pull starts compilation on the same scope on which the compilation of this 
+    * pull (the prefix) ends. The composed pull will end on whatever scope the 
+    * compilation of the `post` pull does.
+    */
+  def flatMap[F2[x] >: F[x], O2 >: O, R2](kont: R => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
+    new Bind[F2, O2, R, R2](this) {
+      def cont(e: Terminal[R]): Pull[F2, O2, R2] =
+        e match {
+          case Succeeded(r) =>
+            try kont(r)
+            catch { case NonFatal(e) => Fail(e) }
+          case res @ Interrupted(_, _) => res
+          case res @ Fail(_)           => res
+        }
+    }
 
-  /** Returns a pull with the result wrapped in `Right`, or an error wrapped in `Left` if the pull has failed. */
-  def attempt: Pull[F, O, Either[Throwable, R]] =
-    map(r => Right(r)).handleErrorWith(t => Succeeded(Left(t)))
+  /** Lazily appends the given `post` pull, to be run after `this` pull.
+    *
+    * - If `this` pull completes successfully, the `post` argument is evaluated to build 
+    *   the pull (this is a lazy parameter), and the compilation starts running the `post` 
+    *   pull. The result of the `post` pull is the result of the composed pull.
+    *   Any carry-over value from `this` pull is discarded.
+    *
+    * - If the compilation of `this` pull is failed or interrupted, then the `post` argument
+    *   is not evaluated and the composed pull returns with the failure or interruption.
+    *
+    * Note that, because the `post` argument is lazy, this method can be used 
+    * to build lazy pulls, in which latest parts of the Pulls are not allocated
+    * into memory until after the first parts have been successfully compiled.
+    *
+    * In both cases, the effectful actions and outputs of the appended pull 
+    * consists of those outputs and actions from the first pull, followed by
+    * those from the `post` pull, in the same order as they would come out of each pull.
+    *
+    * This operation does not add or remove any resource scope boundaries. 
+    * The `post` pull will run on the same scope on which the compilation of 
+    * this pull ends. The composed pull will end on whatever scope the `post` 
+    * pull ended on.
+    */
+  def >>[F2[x] >: F[x], O2 >: O, S](post: => Pull[F2, O2, S]): Pull[F2, O2, S] =
```

Comment:
:-1: on the `post` rename here -- should just be `f`

Reply:
IMHO we should start adding more informative names for the methods, so as to make the code the first documentation. https://hilton.org.uk/blog/naming-smells

Reply:
The `f` name in particular seems to be misleading, seems it is often used for functions and this parameter isn't.

Reply:
Good point about this not being a function -- I'd call it `that`. In general, I'm strongly against trying to create evocative names for function arguments.

Reply:
Can you ellaborate on that preference, or what downsides have you found for "informative" parameter names? 
That is something often recommended by that speaker ([saw a talk at past Scala eXchange](https://skillsmatter.com/skillscasts/11207-how-to-name-things-the-hardest-problem-in-programming)) and other authors such as [Dave Fairley](https://www.youtube.com/watch?v=1KeJc6V4Jjk.)

Reply:
Some miscellaneous conventions:
- `f`, `g`, `h` for function/method parameters of type `FunctionX[...]`
- `A`, `B`, `C` for proper type params
- `F`, `G`, `H` for higher kinded type params
- `A2`, `B2` for type params used for variance reasons (e.g. `def foo[A2 >: A](...)`)

I find small functions, especially when using parametric polymorphism, to be easier to read with such identifiers. There's some commentary about this in FPiS. Considering `def map[B](f: A => B): F[B]` versus `def map[B](convert: A => B): F[B]`, I find the second both harder to write, as we need to come up with a meaningful name for the function, and harder to read, as the `convert` name doesn't stand out as much in the function body.

Consistency is very important too. In general, the codebase follows the conventions laid out above (which we should document somewhere!). Adopting a new style incrementally is jarring.